File: xim/resource/table/EquipmentModelTable.kt
package xim.resource.table

import xim.poc.ItemModelSlot
import xim.poc.RaceGenderConfig
import xim.resource.ByteReader
import xim.util.OnceLogger

private data class TableEntry(val tableOffset: Int, val entryCount: Int)

private data class RaceGenderTable(val entries: Map<ItemModelSlot, List<TableEntry>>)

object EquipmentModelTable {

    private val tables: Map<RaceGenderConfig, RaceGenderTable> by lazy { parse(MainDll.getEquipmentLookupTable()) }

    fun getItemModelPath(raceGenderConfig: RaceGenderConfig, itemModelSlot: ItemModelSlot, itemModelId: Int): String? {
        val table = tables[raceGenderConfig] ?: return null
        val subTable = table.entries[itemModelSlot] ?: return null

        var cumulativeEntryCount = 0

        for (subTableEntry in subTable) {
            if (itemModelId >= cumulativeEntryCount + subTableEntry.entryCount ) {
                cumulativeEntryCount += subTableEntry.entryCount
                continue
            }

            val offset = subTableEntry.tableOffset + (itemModelId - cumulativeEntryCount)
            return FileTableManager.getFilePath(offset)
                .also { OnceLogger.info("[${raceGenderConfig.name}][$itemModelSlot][$itemModelId] -> $it") }
        }

        OnceLogger.error("Failed to evaluate: [$raceGenderConfig] -> [$itemModelSlot] -> [$itemModelId]")
        return null
    }

    fun getNumEntries(raceGenderConfig: RaceGenderConfig, itemModelSlot: ItemModelSlot): Int {
        val table = tables[raceGenderConfig] ?: return 0
        val subTable = table.entries[itemModelSlot] ?: return 0
        return subTable.sumOf { it.entryCount }
    }

    private fun parse(byteReader: ByteReader): Map<RaceGenderConfig, RaceGenderTable> {
        val tables = HashMap<RaceGenderConfig, RaceGenderTable>()

        for (config in RaceGenderConfig.values()) {
            byteReader.position = 0x1B0 * (config.equipmentTableIndex - 1)
            tables[config] = parseRaceGenderTable(byteReader)
        }

        return tables
    }

    private fun parseRaceGenderTable(byteReader: ByteReader): RaceGenderTable {
        val basePosition = byteReader.position
        val table = HashMap<ItemModelSlot, List<TableEntry>>()

        for (slot in ItemModelSlot.values()) {
            byteReader.position = basePosition + 0x30 * (slot.prefix shr 0xC)
            val entries = ArrayList<TableEntry>(6)

            for (i in 0 until 6) {
                val entry = TableEntry(tableOffset = byteReader.next32(), entryCount = byteReader.next32())
                if (entry.tableOffset == 0x0) { continue }
                entries += entry
            }

            table[slot] = entries
        }

        return RaceGenderTable(entries = table)
    }

}

File: lib/resource/table/equipmentModelTable.ts
import { ByteReader } from '../byteReader'
import type { ItemModelSlot, RaceGenderConfig } from './tableTypes'
import { ItemModelSlot as ItemModelSlots, RaceGenderConfig as RaceGenderConfigs } from './tableTypes'

interface TableEntry {
  readonly tableOffset: number
  readonly entryCount: number
}

interface RaceGenderTable {
  readonly entries: ReadonlyMap<ItemModelSlot, readonly TableEntry[]>
}

interface MainDllLike {
  getEquipmentLookupTable(): ByteReader
}

interface FileTableLike {
  getFilePath(fileId: number | null): string | null
}

export class EquipmentModelTable {
  private readonly tables: ReadonlyMap<RaceGenderConfig, RaceGenderTable>

  constructor(mainDll: MainDllLike, private readonly fileTableManager: FileTableLike) {
    this.tables = this.parse(mainDll.getEquipmentLookupTable())
  }

  getItemModelPath(raceGenderConfig: RaceGenderConfig, itemModelSlot: ItemModelSlot, itemModelId: number): string | null {
    const table = this.tables.get(raceGenderConfig)
    if (!table) {
      return null
    }

    const subTable = table.entries.get(itemModelSlot)
    if (!subTable) {
      return null
    }

    let cumulativeEntryCount = 0

    for (const subTableEntry of subTable) {
      if (itemModelId >= cumulativeEntryCount + subTableEntry.entryCount) {
        cumulativeEntryCount += subTableEntry.entryCount
        continue
      }

      const offset = subTableEntry.tableOffset + (itemModelId - cumulativeEntryCount)
      return this.fileTableManager.getFilePath(offset)
    }

    return null
  }

  getNumEntries(raceGenderConfig: RaceGenderConfig, itemModelSlot: ItemModelSlot): number {
    const table = this.tables.get(raceGenderConfig)
    const subTable = table?.entries.get(itemModelSlot)
    return (subTable ?? []).reduce((sum, entry) => sum + entry.entryCount, 0)
  }

  private parse(byteReader: ByteReader): ReadonlyMap<RaceGenderConfig, RaceGenderTable> {
    const tables = new Map<RaceGenderConfig, RaceGenderTable>()

    for (const config of RaceGenderConfigs.values()) {
      byteReader.position = 0x1b0 * (config.equipmentTableIndex - 1)
      tables.set(config, this.parseRaceGenderTable(byteReader))
    }

    return tables
  }

  private parseRaceGenderTable(byteReader: ByteReader): RaceGenderTable {
    const basePosition = byteReader.position
    const table = new Map<ItemModelSlot, TableEntry[]>()

    for (const slot of ItemModelSlots.values()) {
      byteReader.position = basePosition + 0x30 * (slot.prefix >>> 0x0c)
      const entries: TableEntry[] = []

      for (let i = 0; i < 6; i += 1) {
        const entry: TableEntry = {
          tableOffset: byteReader.next32(),
          entryCount: byteReader.next32(),
        }

        if (entry.tableOffset !== 0) {
          entries.push(entry)
        }
      }

      table.set(slot, entries)
    }

    return { entries: table }
  }
}
